src/API/Messages.hs

```diff
@@ -5,26 +5,157 @@ import Types
 import qualified Network.Mail.Parse.Types as MPT
 import Servant
 import Data.Aeson (ToJSON, decode)
+import Data.Aeson.Types (toJSON)
+import qualified Data.Aeson.Types as JSON
 import Database.PostgreSQL.Simple
-import Database.PostgreSQL.Simple.FromRow
+import Database.PostgreSQL.Simple.SqlQQ
 import GHC.Generics
 import Control.Monad.IO.Class (liftIO)
 import Control.Applicative
+import Data.Time.LocalTime (ZonedTime)
+import Data.UUID.Types (UUID)
+import qualified Data.UUID as UUID
+import Data.Maybe (catMaybes, fromJust, fromMaybe, isNothing, isJust)
 
-type MessagesAPI = "messages" :> Get '[JSON] [Message]
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Map.Strict as M
+import Data.Tuple.HT (uncurry3)
 
-data Message = Message {
-  uid :: Int,
-  emailMessage :: Maybe MPT.EmailMessage
+type MessagesAPI = "messages" :> Get '[JSON] [MessageDigest]
+                 :<|> "messages" :> Capture "id" String :> Get '[JSON] MessageDigest
+
+type EmailRelationMap = M.Map UUID [EmailWithRelation]
+type EmailMap = M.Map UUID MPT.EmailAddress
+type ReferenceMap = M.Map UUID [T.Text]
+
+data MessageDigest = MessageDigest {
+  msgId :: !UUID
+, uid :: !Int
+, fromAddr :: Maybe MPT.EmailAddress
+, sentDate :: Maybe ZonedTime
+, replyTo :: Maybe MPT.EmailAddress
+, messageId :: Maybe T.Text
+, references :: [T.Text]
+, inReplyTo :: Maybe T.Text
+, subject :: Maybe T.Text
+, to :: [MPT.EmailAddress]
+, cc :: [MPT.EmailAddress]
+, bcc :: [MPT.EmailAddress]
+, message :: JSON.Value
```

Comment:
I'm not convinced here, is there a way of having distinct types for the collection of messages and a single message without retyping the whole thing? Right now when we're returning a collection of messages, `message` is set to null by the DB